pkg/controller/nodes/dynamic/dynamic_workflow.go

```diff
@@ -38,18 +38,38 @@ func (d dynamicNodeTaskNodeHandler) buildDynamicWorkflowTemplate(ctx context.Con
 	}
 
 	currentAttemptStr := strconv.Itoa(int(nCtx.CurrentAttempt()))
+	uniqueParentID := nCtx.NodeStatus().GetUniqueNodeID()
+
 	// Modify node IDs to include lineage, the entire system assumes node IDs are unique per parent WF.
 	// We keep track of the original node ids because that's where inputs are written to.
 	parentNodeID := nCtx.NodeID()
 	for _, n := range djSpec.Nodes {
+		var subNodeStatus v1alpha1.ExecutableNodeStatus
 		newID, err := hierarchicalNodeID(parentNodeID, currentAttemptStr, n.Id)
 		if err != nil {
 			return nil, err
 		}
-
-		// Instantiate a nodeStatus using the modified name but set its data directory using the original name.
-		subNodeStatus := parentNodeStatus.GetNodeExecutionStatus(ctx, newID)
-
+		if nCtx.ExecutionContext().GetEventVersion() == v1alpha1.V0 {
+			subNodeStatus, err = parentNodeStatus.GetNodeExecutionStatus(ctx, newID)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			// In the newer version we create nodestatus for nodes with the same id
+			// A node execution is uniquely identified with the unique id and not the node_id
+			subNodeStatus, err = parentNodeStatus.GetNodeExecutionStatus(ctx, n.Id)
```

Comment:
Should we update `GetNodeExecutionStatus` to take in the `currentAttemptStr` as well and return a well formed `subNodeStatus` so you don't have to call `Set*` funcs on it after it's returned? it should have all the information it needs to compute that... right?

Reply:
True. It's not just attempt. There are several things. But main reason, they don't apply to majority of the cases today. But let's chat about that.